### PR TITLE
ANDAUTH-57.

### PR DESCRIPTION
**Pull request by jaindiv26**
Added support for OIDC (OpenId connect) into XWiki Authenticator. So this how it works. Suppose ''A' app wants to access the XWiki user info. So this app will start the intent like this -> 

Intent i = new Intent();
i.setComponent(new ComponentName("org.xwiki.android.sync", "org.xwiki.android.sync.activities.OIDC.OIDCActivity"));
startActivityForResult(i, 1);

 

Then XWiki Authenticator activity will open, the user will select the account from the list and upon clicking he'll be taken to the browser and XWiki login page will open. Here the user will enter his credentials and as soon as the login completes, the user will land on consent page to finally give permission to the app. Finally requesting app will get the token. Here the user will get token in the onActivityResult() method and the token lies in the intent data. ->

data.getExtras().get(AccountManager.KEY_AUTHTOKEN).toString();

 

Now the "A" app has the access token.

Issue: - https://jira.xwiki.org/projects/ANDAUTH/issues/ANDAUTH-57?filter=allopenissues

Screenshot_2019-08-09-05-43-12-730
Screenshot_2019-08-09-05-43-28-815

### CODE REVIEW

File: app/src/main/java/org/xwiki/android/sync/AppContext.kt [hunk 1]
```diff
@@ -60,6 +60,10 @@ lateinit var userAccountsCookiesRepo: UserAccountsCookiesRepository
 
 private val apiManagers: MutableMap<UserAccountId, BaseApiManager> = mutableMapOf()
 
+var selectedAccount = ""
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/AppContext.kt]: Again you are using global variables. Clear it, please.

Comment (InsanusMokrassar) [reply app/src/main/java/org/xwiki/android/sync/AppContext.kt]: You must replace it to another place and make it local

File: app/src/main/java/org/xwiki/android/sync/activities/OIDC/OIDCActivity.kt [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+
+        val adapter = OIDCAccountAdapter(this, availableAccountsList, this)
+        binding.lvSelectAccount.adapter = adapter
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        if (requestCode == REQUEST_ACCESS_TOKEN) {
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/activities/OIDC/OIDCActivity.kt]: Replace with `when`:

```
when (requestCode) {
    REQUEST_ACCESS_TOKEN -> { ... }
    REQUEST_NEW_ACCOUNT -> { ... }
}
```

File: app/src/main/java/org/xwiki/android/sync/activities/OIDC/OIDCAuthenticatorActivity.kt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+class OIDCAuthenticatorActivity: AppCompatActivity() {
+
+    private lateinit var flow: AuthorizationCodeFlow
+
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        try {
+            flow = AuthorizationCodeFlow.Builder(
```
Comment (InsanusMokrassar) [inline app/src/main/java/org/xwiki/android/sync/activities/OIDC/OIDCAuthenticatorActivity.kt]: Replace with:

```
AuthorizationCodeFlow.Builder(
    BearerToken.authorizationHeaderAccessMethod(),
    NetHttpTransport(),
    JacksonFactory(),
    GenericUrl(TOKEN_SERVER_URL),
    ClientParametersAuthentication(
        selectedAccount,
        ""
    ),
    selectedAccount,
    AUTHORIZATION_SERVER_URL
).run {
    setScopes(
        mutableListOf(
            "openid",
            "offline_access",
            "profile"
        )
    )
    build()
}
```

**Review by InsanusMokrassar — CHANGES_REQUESTED** [review]
First changes for this PR

File: app/src/main/java/org/xwiki/android/sync/activities/OIDC/OIDCAuthenticatorActivity.kt (hunk 3 above)
Comment (jaindiv26) [reply app/src/main/java/org/xwiki/android/sync/activities/OIDC/OIDCAuthenticatorActivity.kt]: Done.

File: app/src/main/java/org/xwiki/android/sync/activities/OIDC/OIDCActivity.kt (hunk 2 above)
Comment (jaindiv26) [reply app/src/main/java/org/xwiki/android/sync/activities/OIDC/OIDCActivity.kt]: Done.

**Review by InsanusMokrassar — APPROVED** [review]

**Pull request merged**